### ISSUE TITLE
dnsdist: LogAction() log is not flushed on exit

### ISSUE THREAD

**Issue by jpmens** [PowerDNS/pdns#4011]
I have the following:

```
addAction(AllRule(), LogAction("my.log"))
```

When stopping _dnsdist_ with a `shutDown()`, the log is not flushed to disk; the file is created and remains empty after a couple of queries. (I realize a couple of queries isn't _dnsdist_'s main job, but even so it should write the log ;-)

**Comment by rgacogne** [comment]
The `buffered` parameter added in #4329 makes it easier to log a couple of queries, but we still need to flush the log buffer before exiting.

**Comment by rgacogne** [comment]
This behaviour is causing some pain to users again. The issue is that we do not catch the `SIGTERM` signal sent by most process managers, and that POSIX mandates that the default behaviour when that signal is received should be:

> The process is terminated with all the consequences of _exit() except that the status made available to wait() and waitpid() indicates abnormal termination by the specified signal.

Whether output streams will be flushed on `_exit()` is implementation-defined (while it is required on `exit()`, and the glibc's behaviour is to not flush them:

> All open file descriptors in the process are closed. See Low-Level I/O. Note that streams are not flushed automatically when the process terminates; see I/O on Streams.

We could register a `SIGTERM` handler that would call a `beforeExit()` function before calling `_exit()` itself. The `beforeExit()` function would iterate through all our rules/actions, calling some sort of `cleanup()` method that would do nothing by default but would flush the output buffer for `LogAction()` and friends. I guess we should also call that function when the `shutdown()` command is issued via the console.

**Comment by pieterlexis** [comment]
For bonus points, call `sd_notify` with "STOPPING=1" in that function as well [1](https://www.freedesktop.org/software/systemd/man/sd_notify.html#STOPPING=1).

**Comment by rgacogne** [comment]
We need to be careful about what we call from a signal handler, though, only async-signal-safe functions are allowed.

**Comment by omoerbeek** [comment]
The recursor does orderly shutdown by setting a (volatile sig_atomic_t) stop variable which is checked by the threads and then joining the threads. The various destructors take case of cleaning up. No idea if this approach is feasible for dnsdist.

**Comment by rgacogne** [comment]
I don't think we should join the threads because most of them are blocking without timeout (and I don't like to rely on `EINTR`), but setting a flag in the signal handler that is later acted upon in `maintenance` could be an option.
